[DirectX] Set Shader Flag DisableOptimizations.

bharadwajy opened the pull request:
Disabling optimizations using the option `-Od` to `clang-dxc` results in generation of named metadata "dx.disable_optimizations" - as tested [here](https://github.com/llvm/llvm-project/blob/main/clang/test/CodeGenHLSL/disable_opt.hlsl).

- Following changes facilitate setting of Shader Flag `DisableOptimizations`.

  - Detect metadata "dx.disable_optimizations" as part of Metadata Analysis pass and set a new `bool` field 
   `ModuleMetadatainfo::DisableOptimization` accordingly.
  - Add `DXILMetadataAnalysis` as required pass for the pass `DXILShaderFlags` that sets shader flags mask.
  - Set shader flag `DisableOptimizations` based on the value of `ModuleMetadatainfo::DisableOptimizations`.

- Added test to verify setting of shader flag `DisableOptimizations`.
- Updated `llc-pipeline.ll` to reflect changes in output resulting from addition of Metadata Analysis pass as required by `DXILShaderFlags` pass.

NOTE: An [alternative implementation](https://github.com/llvm/llvm-project/compare/main...bharadwajy:llvm-project:shader-flags/disable-opt-flag) that 
  1. does not detect presence of metadata "dx.disable_optimizations" in Metadata Analysis pass and thus does not add the pass as a pre-requisite pass for `DXILShaderFlags` but
  2. detects presence of metadata "dx.disable_optimizations" directly in `DXILShaderFlags` pass

has been considered. However, Metadata Analysis pass seemed appropriate for detection of metadata relevant to setting the shader flag `DisableOptimizations`. Hence this PR.

Closes #112263

Commit:
[DirectX] Set Shader Flag DisableOptimizations.

Disabling optimizations using the option -Od to clang-dxc
results in generation of named metadata "dx.disable_optimizations".

- Detect metadata "dx.disable_optimizations" as part of Metadata
  Analysis pass and set a new bool field
  ModuleMetadatainfo::DisableOptimization accordingly.
- Add DXILMetadataAnalysis as required pass for the pass ShaderFlags
  that sets shader flags mask.
- Set shader flag DisableOptimization based on the value of
  ModuleMetadatainfo::DisableOptimization.

Updated llc-pipeline.ll to reflect changes in output resulting
from addition of Metadata Analysis pass as required by ShaderFlags pass.

bogner commented:
Is the `dx.disable_optimizations` metadata documented somewhere? This all looks like it will work, but I don't really understand why we have a custom metadata node coming from the frontend here rather than just looking at `llvm::Attribute::OptimizeNone` on the entry function or something like that.

bharadwajy commented:
> Is the `dx.disable_optimizations` metadata documented somewhere? This all looks like it will work, but I don't really understand why we have a custom metadata node coming from the frontend here rather than just looking at `llvm::Attribute::OptimizeNone` on the entry function or something like that.

I am not aware of module-level metadata `dx.disable_optimizations` being documented anywhere nor other module-level metadata presently in use - such as `dx.valver`, `dx.shaderModel`. Created [an issue](#124310) to add documentation. 

Currently, presence of command-line option `-Od` is denoted by [adding a module metadata](https://github.com/llvm/llvm-project/blob/7842374103b26933d71a8fe354cd4d8715d55b1c/clang/lib/CodeGen/CGHLSLRuntime.cpp#L189) `dx.disable_optimizations`.  I do not know of the context of the design choice, but it seems to be a reasonable choice to capture the module-wide property of optimizations being disabled. Relying on an attribute of a specific function(s) (such as entry) appears to involve additional bookkeeping or traversals and an awkward way to denote a module-wide property.

bogner commented:
> I am not aware of module-level metadata `dx.disable_optimizations` being documented anywhere nor of other module-level metadata presently in use - such as `dx.valver`, `dx.shaderModel`. Created [an issue](#124310) to add documentation.
>
> Currently, presence of command-line option `-Od` is denoted by [adding a module metadata](https://github.com/llvm/llvm-project/blob/7842374103b26933d71a8fe354cd4d8715d55b1c/clang/lib/CodeGen/CGHLSLRuntime.cpp#L189) `dx.disable_optimizations`. I do not know of the context of the design choice, but it seems to be a reasonable choice to capture the module-wide property of optimizations being disabled. Relying on an attribute of a specific function(s) (such as entry) appears to involve additional bookkeeping or traversals and an awkward way to denote a module-wide property.

There's a very big difference between `dx.disable_optimizations` and metadata like `dx.valver` and `dx.shaderModel` - the latter are part of DXIL and we're required to generate them to match dxc/fxc output. `dx.disable_optimizations` is unique to the upstream llvm implementation (it was added in f712c0131f3b without any specific mention of why it was needed). My concern here is mostly that the `dx.disable_optimizations` metadata is redundant with information that's already available in more standard ways and the story around keeping it consistent within the module isn't clear. It also isn't clear to me how this will interact with linking - what if part of the module is built with optimizations disabled and part with them enabled? Reconciling global metadata in this case sounds like a nightmare, whereas the `optnone` attribute that is already present on functions will already do the right thing.

Commit:
Delete metadata dx.disable_optimizations, if it exists,
from module after its presence has been noted in
ModuleMetadata Analysis pass. This ensures that this
metadata that is not part of DXIL specification
is not emitted in the final output.

Update disable-opt.ll to verify dx.disable_optimizations
metadata is deleted.

bharadwajy commented:
Since `optnone` function atribute is not set when `-O0` (or `-Od`) is specified, my experiements to set them appear to reveal that such a change interferes adversely with the exisiting DXIL linkage assumptions / rules. To prevent "non-standard" DXIL metadata (i.e., that not specified in DXIL Spec) from being generated in the final DXIL output, I pushed changes to delete the metadata once its intent is captured in Metadata analysis pass. Following the discussion offline, @bogner, please see if this would address the concerns raised. Thanks!

bharadwajy commented:
>  [...] It also isn't clear to me how this will interact with linking - what if part of the module is built with optimizations disabled and part with them enabled? Reconciling global metadata in this case sounds like a nightmare, whereas the `optnone` attribute that is already present on functions will already do the right thing.

With the changes that delete the metadata `dx.disable_optimizations`, the only thing a DXIL linker would rely on to know if optimizations were disabled during compilation of modules being linked is the documented (and generated) shader flag `DisableOptimizations`. I believe DXC linking rules have a (some?) solution for this or we may have to revisit them for the ongoing HLSL support for HLSL, irrespective of the mechanism to set the shader flag `DisableOptimizations`.

hekota reviewed (approved):
LGTM, though I also find it strange that there is no module-level flag in Clang which signifies that optimizations are disabled and that we have to rely on `dx.disable_optimizations`.

llvm-beanz commented on llvm/lib/Analysis/DXILMetadataAnalysis.cpp:
```diff
@@ -68,6 +68,37 @@ static ModuleMetadataInfo collectMetadataInfo(Module &M) {
+  // emitted in the final DXIL output, now that its intent is captured in MMDAI,
+  // if present.
+
+  if (NewFlagEntries.size() != FlagEntries.size()) {
```
Do we need to do this here? We have code in DXILPrepare that cleans the module flags.

@bogner, I think the reason the `dx.disable_optimizations` named metadata wasn't causing issues is because DXIL doesn't use custom module flags at all, so in DXILPrepare we just strip out any non-standard LLVM 3.7 module flags, which will clean this up.

bharadwajy replied:
> Do we need to do this here? We have code in DXILPrepare that cleans the module flags.
> 
> @bogner, I think the reason the `dx.disable_optimizations` named metadata wasn't causing issues is because DXIL doesn't use custom module flags at all, so in DXILPrepare we just strip out any non-standard LLVM 3.7 module flags, which will clean this up.

@llvm-beanz I [proposed a PR](#125937) to decorate shader entry functions with the attribute `optnone` as an alternative to using `dx.disable_optimizations` named metadata - per feedback [above](#issuecomment-2611684827). 

The idea is to eliminate creation of `dx.disable_optimizations` metadata and key off the `optnone` attribute on shader entry functions to set the shader flag `DisableOptimizations` as implemented [here](https://github.com/llvm/llvm-project/compare/main...bharadwajy:llvm-project:delete-disable-opt-metadata?expand=1) with an expectation of converting it to a follow-on PR to #125937.

I'd appreciate a feedback on #125937 in this context as well as the preferred choice compared to this PR.

llvm-beanz replied:
This looks like it now needs an update to adjust to #125937 landing.

bogner commented:
This is superceded by #126813, right? Please close this if it's no longer relevant.

bharadwajy commented:
Closing. Superceded by #126813.
